src/storage/hummock_sdk/src/filter_key_extractor.rs

```diff
@@ -152,22 +155,26 @@ impl FilterKeyExtractor for SchemaFilterKeyExtractor {
         let (_vnode_prefix, pk) = key.split_at(VirtualNode::SIZE);
 
         // if the key with table_id deserializer fail from schema, that should panic here for early
-        // detection
+        // detection.
         let pk_prefix_len = self
             .deserializer
-            .deserialize_prefix_len_with_column_indices(pk, 0..self.read_pattern_prefix_column)
+            .deserialize_prefix_len_with_column_indices(
+                pk,
+                self.distribution_key_start_index_in_pk..self.read_pattern_prefix_column,
+            )
             .unwrap();
 
         let prefix_len = TABLE_PREFIX_LEN + VirtualNode::SIZE + pk_prefix_len;
-        &full_key[0..prefix_len]
+        &full_key[self.distribution_key_start_index_in_pk..prefix_len]
```

Comment:
Only modifying the implementation of FilterKeyExtractor to take distribution_key_start_index into consideration is not enough since its users may also assume that the extracted key is a prefix of the full key. Example: https://github.com/Little-Wallace/risingwave/blob/e2191490b82af6b1dd6d42c684a4c6c3c3ae29d8/src/storage/src/hummock/sstable/builder.rs#L205

Please review the usages of FilterKeyExtractor to see how the extracted key is used.

Reply:
If we break this assumption, we only need to maintain `start_index` more in order to get `last_extract_key`